pkg/executor/join/hash_join_v2.go

```diff
@@ -401,184 +376,145 @@ func (e *HashJoinV2Exec) initializeForProbe() {
-
+func (e *HashJoinV2Exec) startProbeWorkers(ctx context.Context, fetcherAndWorkerSyncer *sync.WaitGroup) {
+	fetcherAndWorkerSyncer.Add(int(e.hashJoinCtxBase.Concurrency))
 	for i := uint(0); i < e.Concurrency; i++ {
 		workerID := i
-		e.workerWg.RunWithRecover(func() {
-			defer trace.StartRegion(ctx, "HashJoinWorker").End()
-			e.ProbeWorkers[workerID].runJoinWorker()
-		}, e.ProbeWorkers[workerID].handleProbeWorkerPanic)
+		e.waiterWg.RunWithRecover(
+			func() {
+				defer trace.StartRegion(ctx, "HashJoinWorker").End()
+				e.ProbeWorkers[workerID].runJoinWorker()
+			},
+			func(r any) {
+				handleError(e.joinResultCh, &e.finished, r)
+				fetcherAndWorkerSyncer.Done()
+			},
+		)
 	}
-	e.waiterWg.RunWithRecover(e.waitJoinWorkersAndCloseResultChan, nil)
 }
 
-func (w *ProbeWorkerV2) handleProbeWorkerPanic(r any) {
-	if r != nil {
-		w.HashJoinCtx.joinResultCh <- &hashjoinWorkerResult{err: util.GetRecoverError(r)}
-	}
+func (e *HashJoinV2Exec) startFinalWorker(syncer chan struct{}) {
+	e.waiterWg.RunWithRecover(
+		func() {
+			e.finalWorker(syncer)
+		},
+		func(r any) {
+			handleError(e.joinResultCh, &e.finished, r)
+		},
+	)
 }
 
-func (e *HashJoinV2Exec) handleJoinWorkerPanic(r any) {
-	if r != nil {
-		e.joinResultCh <- &hashjoinWorkerResult{err: util.GetRecoverError(r)}
-	}
+func (e *HashJoinV2Exec) startProbeFetcher(ctx context.Context) {
+	// Fetcher needs to wake up finalWorker after all probe workers finish tasks.
+	// This syncer could let fetcher knows if probe workers have finished their tasks.
+	fetcherAndWorkerSyncer := &sync.WaitGroup{}
+
+	// synchronize between probe fetcher and final worker
+	pfAndFWSync := make(chan struct{}, 1)
+
+	e.startProbeWorkers(ctx, fetcherAndWorkerSyncer)
+	e.startFinalWorker(pfAndFWSync)
+
+	defer func() {
+		if r := recover(); r != nil {
+			handleError(e.joinResultCh, &e.finished, r)
+		}
+
+		close(pfAndFWSync)
+	}()
+
+	e.ProbeSideTupleFetcher.fetchProbeSideChunks(
+		ctx,
+		e.MaxChunkSize(),
+		func() bool { return e.ProbeSideTupleFetcher.hashTableContext.hashTable.isHashTableEmpty() },
+		e.ProbeSideTupleFetcher.canSkipProbeIfHashTableIsEmpty,
+		e.ProbeSideTupleFetcher.needScanRowTableAfterProbeDone,
+		e.ProbeSideTupleFetcher.shouldLimitProbeFetchSize(),
+		&e.ProbeSideTupleFetcher.hashJoinCtxBase)
+
+	e.ProbeSideTupleFetcher.closeProbeResultChs()
+
+	fetcherAndWorkerSyncer.Wait()
+
+	// Wake up final worker
+	pfAndFWSync <- struct{}{}
+}
+
+func (e *HashJoinV2Exec) fetchAndProbeHashTable(ctx context.Context) {
+	e.waiterWg.RunWithRecover(
+		func() {
+			defer trace.StartRegion(ctx, "HashJoinProbeSideFetcher").End()
+			e.startProbeFetcher(ctx)
+		},
+		func(r any) {
+			handleError(e.joinResultCh, &e.finished, r)
+		},
+	)
 }
 
-func (e *HashJoinV2Exec) waitJoinWorkersAndCloseResultChan() {
-	e.workerWg.Wait()
+// finaWorker is responsible for scanning the row table after probe done and wake up the build fetcher
+func (e *HashJoinV2Exec) finalWorker(syncer chan struct{}) {
```

Comment:
I think if the join don't need to scan hash table after probe, there is no need to start final worker?

Reply:
> I think if the join don't need to scan hash table after probe, there is no need to start final worker?

Yes, but final worker will not create more go routines when there is no need to scan hash table.